File: api/crs/codegen/lib/crs_solve_triu/crs_solve_triu.c
#include "crs_solve_triu.h"

void crs_solve_triu(const CRS_Matrix *A, emxArray_real_T *b)
{
  int n;
  int i;
  int b_i;
  int i1;
  int i2;
  int cind;
  n = A->row_ptr->size[0] - 1;
  for (i = n; i >= 1; i--) {
    b_i = A->row_ptr->data[i - 1];
    i1 = b_i + 1;
    i2 = A->row_ptr->data[i] - 1;
    for (cind = i1; cind <= i2; cind++) {
      b->data[i - 1] -= A->val->data[cind - 1] * b->data[A->col_ind->data[cind -
        1] - 1];
    }

    b->data[i - 1] /= A->val->data[b_i - 1];
  }
}

void crs_solve_triu_initialize(void)
{
}

void crs_solve_triu_terminate(void)
{
}

File: api/crs/crs_solve_triu.m
function b = crs_solve_triu(A, b)
% crs_solve_triu Solves A\b, where A is upper triangular
%     b = crs_solve_triu(A, b)
%  The matrix is assumed to be upper triangular. The right-hand side vector
%  will be overwritten by A\b at output
%
% See also crs_solve_tril, crs_solve_trilt

%#codegen -args {crs_matrix, coder.typeof(0, [inf,1])}

for i = int32(length(A.row_ptr)) - 1:-1:1
    rstart = A.row_ptr(i);
    assert(A.col_ind(rstart) == i && A.val(rstart) ~= 0);

    for cind = (rstart + 1):(A.row_ptr(i+1) - 1)
        k = A.col_ind(cind);
        b(i) = b(i) - A.val(cind) * b(k);
    end

    b(i) = b(i) / A.val(rstart);
end


function test %#ok<DEFNU>
%!test
%! for k=1:100
%!     R = triu(rand(10,10)); b = rand(10,1);
%!     if condest(R) < 1.e5
%!         A = crs_createFromSparse(sparse(R));
%!         us = crs_solve_triu(A, b);
%!         assert(norm(R * us - b) <= 1.e-10);
%!     end
%! end
